pkgs/tools/security/sha1collisiondetection/default.nix

```diff
@@ -0,0 +1,39 @@
+{ stdenv, fetchFromGitHub, libtool }:
+
+stdenv.mkDerivation  rec {
+  pname = "sha1collisiondetection";
+  version = "2017-02-21";
+  name = "${pname}-unstable-${version}";
+
+  src = fetchFromGitHub {
+    owner = "cr-marcstevens";
+    repo = "${pname}";
+    rev = "40ccf5e3537ff6b2e5ceac5747f376a2ef430bec";
+    sha256 = "1k2qjarcx88ndc4yl8xnr1l3j74w1qacvyrhv4ms9hndy8lrsm0l";
+  };
+
+  patchPhase = ''substituteInPlace Makefile --replace "shell arch" "uname -m"'';
+
+  buildInputs = [ libtool ];
+
+  installPhase = ''
+    mkdir -p $out/bin $out/lib
+    cp bin/sha1dcsum bin/sha1dcsum_partialcoll $out/bin
+    cp bin/libdetectcoll.la $out/lib/
+    '';
+
+  meta = with stdenv.lib; {
+    description = "Library and command line tool to detect SHA-1 collision";
+    longDescription = ''
+      This library and command line tool were designed as near drop-in
+      replacements for common SHA-1 libraries and sha1sum. They will
+      compute the SHA-1 hash of any given file and additionally will
+      detect cryptanalytic collision attacks against SHA-1 present in
+      each file. It is very fast and takes less than twice the amount
```

Comment:
Why run any risk, if the solution is superior in every respect other than its age?

You are right that this probably needs a significant effort, but this scientific scoop has been in the works for years and with collaborators outside of science. Like the Kaminsky DNS hack, it now comes down to ingenuity and time - and with high stakes, both are available to quite some shady organisations. File archives are within direct scope:

> Our work shows that it is now practical to find collisions for SHA1 and that thus it is not secure to use for digital signatures, file integrity, and file identification purposes.

The author said this in an interview:
 
 > As an example of the insecurity of SHA1 [the identical-prefix collision] is very meaningful. Chosen-prefix collisions are more powerful and relieve an attacker from some constraints, but identical-prefix collisions can be just as dangerous with a very creative attacker. It is hard to say which things you can’t do with an identical-prefix collision.

> But we do know quite some demonstrated threats using identical-prefix collisions: certificates with identical names and different pubkeys, PDF files, PostScript files, TIFF files, JPEG files, Word files, file archives, signed software, Email message/attachment PGP/GPG signatures.

I don't know what Nix users use sha1 hashes for, but from my perspective there seem to be no reasons to hang on to the broken version as the new one is:

a) more future-proof against known and suspected attacks 
b) faster (so it would speed up Nix)
c) otherwise functionally identical

(the first thing I would do would be to look for near-collision blocks around all the popular open source license texts - that would open up a lot of hurt)